sagemaker-debugger/build_your_own_container_with_debugger/docker/tf_keras_resnet_byoc.py

```diff
@@ -0,0 +1,110 @@
+"""
+This script is a ResNet training script which uses Tensorflow's Keras interface, and acts as an
+example of how to use SageMaker Debugger if you use either your own custom container to run in SageMaker, or your own script outside SageMaker.
+It has been orchestrated with SageMaker Debugger hooks to allow saving tensors during training.
+These hooks have been instrumented to read from json configuration that SageMaker will put in the training container.
+Configuration provided to the SageMaker python SDK when creating a job will be passed on to the hook.
+This allows you to use the same script with differing configurations across different runs.
+
+If you use an official SageMaker Framework container (i.e. AWS Deep Learning Container), then
+you do not have to orchestrate your script as below. Hooks will automatically be added in those environments. This experience is termed as "Zero Script Change". For more information, please refer to https://github.com/awslabs/sagemaker-debugger/blob/master/docs/sagemaker.md#zero-script-change. An example of the same is provided at https://github.com/awslabs/amazon-sagemaker-examples/sagemaker-debugger/tensorflow2/tensorflow2_zero_code_change.
+"""
+
+# Standard Library
+import argparse
+import random
+
+# Third Party
+import numpy as np
+import tensorflow.compat.v2 as tf
+from tensorflow.keras.applications.resnet50 import ResNet50
+from tensorflow.keras.datasets import cifar10
+from tensorflow.keras.utils import to_categorical
+
+# smdebug modification: Import smdebug support for Tensorflow
+import smdebug.tensorflow as smd
+
+
+def train(batch_size, epoch, model, hook):
+    (X_train, y_train), (X_valid, y_valid) = cifar10.load_data()
+
+    Y_train = to_categorical(y_train, 10)
+    Y_valid = to_categorical(y_valid, 10)
+
+    X_train = X_train.astype('float32')
+    X_valid = X_valid.astype('float32')
+
+    mean_image = np.mean(X_train, axis=0)
+    X_train -= mean_image
+    X_valid -= mean_image
+    X_train /= 128.
+    X_valid /= 128.
+    
+    # register hook to save the following scalar values
+    hook.save_scalar("epoch", epoch)
+    hook.save_scalar("batch_size", batch_size)
+    hook.save_scalar("train_steps_per_epoch", len(X_train)/batch_size)
+    hook.save_scalar("valid_steps_per_epoch", len(X_valid)/batch_size)
+    
+    model.fit(X_train, Y_train,
+              batch_size=batch_size,
+              epochs=epoch,
+              validation_data=(X_valid, Y_valid),
+              shuffle=False,
+              # smdebug modification: Pass the hook as a Keras callback
+              callbacks=[hook])
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Train resnet50 cifar10")
+    parser.add_argument("--batch_size", type=int, default=50)
+    parser.add_argument("--epoch", type=int, default=100)
```

Comment:
<=5 epochs should suffice for this example notebook

Reply:
I would say less than 5 epochs is too short to show the whole scope of training. As suggested in the analysis part, I showed that the overfitting issue is raised at step 7200, which is included in the 6th epoch. I would set it to 10 epochs.